### PR TITLE
Reduce the unnecessary updateStoreStatusLocked call counts.

### PR DESCRIPTION
**Pull request by mantuliu**
### What problem does this PR solve? 
Improve the heartbeat processing
UCP #2395 

### What is changed and how it works?
Reduce the call counts of updateStoreStatusLocked.

The test conditions are: 1 million region 20store 0.1 region-update-ratio, 10heartbeat-rounds. The average heartbeat is the average of the last 9 times.

the improve percent like this:
no.|improve content |related code |average heartbeat/s | improve percent
---|---|---|---|---
1|   master branch |     |   101652 | 0%    
2|use ShallowClone when UpdateStoreStatus.|   store.go 666: newStore := store.ShallowClone(SetLeaderCount(leaderCount),  | 124109   |     22%
3|use reduce the number of updates when update related stores.|   cluster.go 572-583: c.updateStoreStatusLocked(key),  | 132076   |     7.8%

Tests 

 - Unit test

### CODE REVIEW

File: server/cluster/cluster.go [hunk 1]
```diff
@@ -569,13 +569,17 @@ func (c *RaftCluster) processRegionHeartbeat(region *core.RegionInfo) error {
 		}
 
 		// Update related stores.
+		storeMap := make(map[uint64]uint64)
```
Comment (rleungx) [inline server/cluster/cluster.go:572]: How about `storeMap := make(map[uint64]struct{})`?

Comment (mantuliu) [reply server/cluster/cluster.go]: Done. Thanks.

**Comment by mantuliu** [comment]
> Hi, @mantuliu I test with
> 
> ```
>  ./bin/pd-heartbeat-bench -region-update-ratio=0.1 --heartbeat-rounds=10
> ```
> 
> only got
> 
> ```
> Summary:
>   Total:        15.1577 secs.
>   Slowest:      0.3920 secs.
>   Fastest:      0.0000 secs.
>   Average:      0.0003 secs.
>   Stddev:       0.0082 secs.
>   Requests/sec: 65974.2556
> ```
> 
> In my test environment, I do not see about 7.8% imporve vs master, the QPS roughtly same. Can you tell me about your test method?
> BTW, the benchmark only updates region with [increase the version](https://github.com/pingcap/pd/blob/master/tools/pd-heartbeat-bench/main.go#L163-L165), do not change the peer. it's may affect the result.

Hi @nolouch My test like this:
```
var (
	pdAddr            = flag.String("pd", "127.0.0.1:2379", "pd address")
	storeCount        = flag.Int("store", 20, "store count")
	regionCount       = flag.Uint64("region", 1000000, "region count")
	keyLen            = flag.Int("keylen", 56, "key length")
	replica           = flag.Int("replica", 3, "replica count")
	regionUpdateRatio = flag.Float64("region-update-ratio", 0.1, "the ratio of the region need to update")
	sample            = flag.Bool("sample", false, "sample per second")
	heartbeatRounds   = flag.Int("heartbeat-rounds", 10, "the total rounds of hearbeat")
)
```
And I used the pd-heartbeat-bench of the master branch.
Calculation formula: 7.8% = (132076 - 124109)/101652.
My computer hardware like this:
```
macOS 10.15.3
cpu:2.3GHz 8 intel Core i9
memory: 32G 2667 MHz DDR4
storage: Macintosh HD 1T
```

**Comment by mantuliu** [comment]
The result like this:
Summary:
  Total:        7.6203 secs.
  Slowest:      0.7093 secs.
  Fastest:      0.0000 secs.
  Average:      0.0001 secs.
  Stddev:       0.0054 secs.
  Requests/sec: 131231.3358

**Comment by mantuliu** [comment]
> Is this result the same for each round? From my test, there are some errors every time.

Not all rounds are the same in my test environment. What kind of error is it?

**Comment by rleungx** [comment]
I mean the different results from my testing, sometimes the master branch is better.

**Comment by rleungx** [comment]
But from the code perspective, this is indeed an improvement, LGTM.

**Pull request merged**